R/shiny.R

```diff
@@ -962,20 +962,37 @@ ShinySession <- R6Class(
     },
     flushOutput = function() {
 
+      # Return TRUE if there's any stuff to send to the client.
+      hasPendingUpdates <- function() {
+        return(
+          length(private$invalidatedOutputValues) != 0 ||
+          length(private$invalidatedOutputErrors) != 0 ||
+          length(private$inputMessageQueue) != 0
+        )
+      }
+
       # ..stacktraceon matches with the top-level ..stacktraceoff..
       private$flushCallbacks$invoke(..stacktraceon = TRUE)
-      # ..stacktraceon matches with the top-level ..stacktraceoff..
-      on.exit(private$flushedCallbacks$invoke(..stacktraceon = TRUE))
 
-      if (length(private$progressKeys) == 0
```

Comment:
Previously, the check for `private$progressKeys` didn't quite have the right logic, I believe. If there was anything in `private$progressKeys`, *all* of the subsequent code in this function would run, including sending a message that didn't contain anything from `private$progressKeys`:

```R
 private$sendMessage(
         errors = as.list(errors),
         values = as.list(values),
         inputMessages = inputMessages
       )
```

I think that instead, `private$progressKeys` should just get reset every time this function is called. (That's how I've implemented it, below.)